Adding fabric content (#829)
microsoft/finops-toolkit: docs-mslearn/fabric/create-finops-workspace.md

Reviewer: @bandersmsft I renamed this file to "create-finops-workspace.md" to simplify and align with the future naming.

---
title: Create a Fabric workspace for FinOps
description: This article walks you through creating and configuring a Microsoft Fabric workspace for FinOps.
author: bandersmsft
ms.author: banders
ms.date: 07/17/2024
ms.topic: how-to
ms.service: finops
ms.reviewer: micflan
---

# Create a Fabric workspace for FinOps

This article walks you through creating and configuring a [Microsoft Fabric](/fabric/get-started/microsoft-fabric-overview) workspace for FinOps. It provides a step-by-step guide to create and configure Azure Data Lake storage (ADLS) setup, data export, workspace creation, and data ingestion. When completed, you can use Power BI to build reports.

## Prerequisites

Before you begin, you must have:

- A paid Azure subscription with cost and usage information

To complete this walkthrough, you create the following resources that incur costs:

- An Azure Data Lake Storage Gen2 account

> - A Fabric or Power BI workspace

Reviewer: Doesn't this require Fabric explicitly? I don't think PBI only works, right?

Author: Right. I"ve fixed.

> ## Create and configure Azure Data Lake Storage

Reviewer: ```suggestion
<br>

## Create and configure Azure Data Lake Storage
```

- Option 1: To create a Data Lake storage account, see [Create a storage account for Azure Data Lake Storage Gen2](/azure/storage/blobs/create-data-lake-storage-account).
- Option 2: To enable hierarchical namespace on an existing storage account, see [Upgrade Azure Blob Storage with Azure Data Lake Storage Gen2 capabilities](/azure/storage/blobs/upgrade-to-data-lake-storage-gen2-how-to?tabs=azure-portal).

>[!TIP]
>If you deployed [FinOps hubs](https://aka.ms/finops/hubs), you can use the storage account created as part of the hub resource group.

For the following example, we used the following Data Lake Gen 2 storage account. You use your own storage account and subscription details.

- Subscription = *Contoso subscription*.
- Storage account = *Contoso-storage-account*.

:::image type="content" source="./media/create-fabric-workspace-finops/storage-account-configuration-gen-2.png" border="true" alt-text="Screenshot showing converting an existing storage account into Data Lake storage." lightbox="./media/create-fabric-workspace-finops/storage-account-configuration-gen-2.png" :::

> ## Export cost data

Reviewer: ```suggestion
<br>

## Export cost data
```

> You can use existing actual or amortized cost exports; however, the following example creates a new cost and usage details (FOCUS) data export.
> 
> What is FOCUS? FinOps Open Cost and Usage Specification (FOCUS) is a new, cross-cloud billing data standard hosted by the FinOps Foundation and defined by Microsoft and a collection of FinOps practitioners, cloud providers, and vendors. To learn more about FOCUS and its benefits, see [About FOCUS](https://microsoft.github.io/finops-toolkit/focus). The FinOps Foundation also offers a free [Introduction to FOCUS course](https://learn.finops.org/introduction-to-focus). Microsoft’s Power BI solutions for FinOps are aligned to FOCUS.

Reviewer: I want to lean on FOCUS a little more to encourage people to use that rather than actual/amortized.
```suggestion
Since we're creating a workspace optimized for FinOps, we'll export cost data using the FinOps Open Cost and Usage Specification (FOCUS), a provider-agnostic data format for cost details. All our examples will use FOCUS but you can also use existing actual or amortized cost exports that have file overwriting enabled.

To learn more about FOCUS and its benefits, see [About FOCUS](https://microsoft.github.io/finops-toolkit/focus). The FinOps Foundation also offers a free [Introduction to FOCUS course](https://learn.finops.org/introduction-to-focus). Microsoft’s Power BI solutions for FinOps are aligned to FOCUS.
```

To create an export, see [Create exports](/azure/cost-management-billing/costs/tutorial-improved-exports#create-exports).

Here are the high-level steps to create an export:

> 1. To navigate:
>     1. Sign in to the Azure portal at [https://portal.azure.com](https://portal.azure.com/), search for **Cost Management**.
>     2. Select the required scope and select **Exports** in the left navigation menu.
>     3. Select **+ Create**
> 2. On the Basics tab:
>     1. Select the template = **Cost and usage (FOCUS)**
>         :::image type="content" source="./media/create-fabric-workspace-finops/exports-cost-and-usage-focus.png" border="true" alt-text="Screenshot showing creating a new FOCUS export dataset." lightbox="./media/create-fabric-workspace-finops/exports-cost-and-usage-focus.png" :::
> 3. On the Datasets tab:
>     1. Fill in **Export prefix** to ensure you have a unique name for the export. For example, *june2024focus*.
> 4. On the Destination tab:
>     1. Storage type = Azure blob storage
>     2. Destination and storage = **Use existing**
>     3. Subscription = *Contoso subscription*
>     4. Storage account = *Contoso-storage-account*
>     5. Container = *msexports*
>     6. Directory = focuscost/&lt;*scope-id*&gt;
>     7. Overwrite data = **Enabled**
> 5. On the Review + Create tab:
>     - Select **Create**

Reviewer: Simplifying the steps a tad...
```suggestion
1. Sign in to the Azure portal at [https://portal.azure.com](https://portal.azure.com/), search for **Cost Management**.
2. Select the required scope and select **Exports** in the left navigation menu.
3. Select **+ Create**
4. On the Basics tab, select the template = **Cost and usage (FOCUS)**
   :::image type="content" source="./media/create-fabric-workspace-finops/exports-cost-and-usage-focus.png" border="true" alt-text="Screenshot showing creating a new FOCUS export dataset." lightbox="./media/create-fabric-workspace-finops/exports-cost-and-usage-focus.png" :::
5. On the Datasets tab, fill in **Export prefix** to ensure you have a unique name for the export. For example, *june2024focus*.
6. On the Destination tab, select:
   - Storage type = Azure blob storage
   - Destination and storage = **Use existing**
   - Subscription = *Contoso subscription*
   - Storage account = *Contoso-storage-account*
   - Container = *msexports*
   - Directory = focuscost/&lt;*scope-id*&gt;
   - Overwrite data = **Enabled**
7. On the Review + Create tab, select **Create**
```

> Ensure that your export completed and that the file is available before moving to the next step.

Reviewer: Is this referring to the CSV file that was exported? If so, we should probably call that out. Also, exports dont run automatically, so we need to add a step above to run the export after it's created.

Author: gotcha.

>[!IMPORTANT]

> > Cost Management exports utilize managed identity. To create an export, you must be an Owner of the storage account or have access to grant permissions to the storage account. To see all required permissions, see [Prerequisites](/azure/cost-management-billing/costs/tutorial-improved-exports#prerequisites).

Reviewer: Should this be in our prerequisites above?

Author: added to prereqs

If you want to automate export creation, consider using the [New-FinOpsCostExport command](https://microsoft.github.io/finops-toolkit/powershell/cost/New-FinOpsCostExport) in the FinOps toolkit PowerShell module.

> If you deployed [FinOps hubs](https://aka.ms/finops/hubs), you can use the exports configured against the hub.

Reviewer: ```suggestion
If you deployed [FinOps hubs](https://aka.ms/finops/hubs), you can skip this step and use the processed data in hub storage.
```

> ## Create a Fabric workspace

Reviewer: ```suggestion
<br>

## Create a Fabric workspace
```

> ## Create a Fabric workspace

Reviewer: Since we include this in the prerequisites, do we need a section for it? I'm fine either way. Just thinking out loud. If we remove this one, we should probably add the links to create one above in the prereqs section.

NOTE: The rest of my feedback for this section is assuming we keep it.

Author: I think we should keep the section in case people haven't already created a workspace. I intentionally worded the prereq to say:  "To complete this walkthrough, you create the following resources that incur costs: - A Fabric workspace "

1. Sign in to the Fabric app using your [Fabric free account](https://www.microsoft.com/microsoft-fabric/getting-started) or [Microsoft Power BI](https://app.powerbi.com) account.
2. You can create a new workspace for your FinOps resources or choose to use an existing workspace. To create a workspace, see [Create a workspace](/fabric/get-started/create-workspaces).

> If you already have a [Microsoft Fabric workspace](/fabric/get-started/workspaces), you can skip creating a new workspace, unless you want to keep your FinOps resources separate from others.

Reviewer: Should we put this before the steps?

Author: good point. fixed.

> If you didn't set up Microsoft Fabric, see [Enable Microsoft Fabric for your organization](/fabric/admin/fabric-switch).

Reviewer: Should we make this the first step to enable Fabric if it hasn't been enabled?

Author: yep. moved

If you’re new to Microsoft Fabric or don’t have Fabric capacity enabled, see [Fabric trial](/fabric/get-started/fabric-trial). If Fabric trials are disabled, you might need to [create a new Fabric capacity](/fabric/admin/capacity-settings?tabs=fabric-capacity#create-a-new-capacity).

Nontrial Fabric capacity incurs charges for your organization. See [Microsoft Fabric pricing](https://azure.microsoft.com/pricing/details/microsoft-fabric/).

Here's an example screenshot showing a new Fabric workspace getting created.

:::image type="content" source="./media/create-fabric-workspace-finops/fabric-create-workspace.png" border="true" alt-text="Screenshot showing a workspace getting created." lightbox="./media/create-fabric-workspace-finops/fabric-create-workspace.png" :::

> ## Create a lakehouse

Reviewer: ```suggestion
<br>

## Create a lakehouse
```

You can either create a new lakehouse or use an existing one. Lakehouses are interoperable so we would generally recommend using a single lakehouse for all your FinOps datasets. A single lakehouse keeps related information together to share data models, pipelines, and security measures.

To create a new lakehouse in the Fabric workspace of your choice:

> 1. Select **New** > **More options**.  

Reviewer: Do we need to set context on where they go to select this option?

Author: Updated

    :::image type="content" source="./media/create-fabric-workspace-finops/fabric-new-more-options.png" border="true" alt-text="Screenshot showing navigation to More options." lightbox="./media/create-fabric-workspace-finops/fabric-new-more-options.png" :::
3. Select **Lakehouse**  
    :::image type="content" source="./media/create-fabric-workspace-finops/fabric-new-lakehouse.png" border="true" alt-text="Screenshot showing the Lakehouse option." lightbox="./media/create-fabric-workspace-finops/fabric-new-lakehouse.png" :::

For more information, see [Create a lakehouse](/fabric/data-engineering/create-lakehouse).

>[!NOTE]

> >Microsoft Fabric [lakehouses](/fabric/data-engineering/lakehouse-overview) are data architectures that allow organizations to store and manage structured and unstructured data in a single location. They use  various tools and frameworks to process and analyze that data. These tools and frameworks can include SQL-based queries and analytics, machine learning, and other advanced analytics techniques.

Reviewer: nit: Should this be the intro paragraph for this section? I'm fine either way.

Author: Moved to section intro as suggested.

> ## Create a shortcut to storage

Reviewer: ```suggestion
<br>

## Create a shortcut to storage
```

Shortcuts in lakehouse allow you to reference data without copying it. It unifies data from different lakehouses, workspaces, or external storage, such as Data Lake Gen2 or Amazon Web Services (AWS) S3. You can quickly make large amounts of data available in your lakehouse locally without the latency of copying data from the source.

To create a shortcut, see [Create an Azure Data Lake Storage Gen2 shortcut](/fabric/onelake/create-adls-shortcut).

> 1. Select the lakehouse of your choice.

Reviewer: nit: Should we say where to start?

Author: updated

2. Select the ellipsis (**...**) next to **Files**.  
    You add CSV and Parquet files under **Files**. Delta tables get added under **Tables**.
3. Select **New shortcut**.  
    :::image type="content" source="./media/create-fabric-workspace-finops/fabric-new-shortcut.png" border="true" alt-text="Screenshot showing creating a new shortcut in a lakehouse under the Files folder." lightbox="./media/create-fabric-workspace-finops/fabric-new-shortcut.png" :::

> 4. Select **Azure Data Lake Storage Gen 2** and provide the following settings.
>     1. URL = **Data Lake Storage** URL of the Data Lake storage account. See the following note about authentication.
>     2. Connection = **Create a new connection**
>     3. Connection name = &lt;*Any name of your choice*&gt;
>     4. Authentication kind = **Organizational account**
>     5. Sign in when prompted.

Reviewer: ```suggestion
4. Select **Azure Data Lake Storage Gen 2** and provide the following settings:
   - URL = **Data Lake Storage** URL of the Data Lake storage account. See the following note about authentication.
   - Connection = **Create a new connection**
   - Connection name = &lt;*Any name of your choice*&gt;
   - Authentication kind = **Organizational account**
   - Sign in when prompted.
```

Here’s an example screenshot showing the New shortcut connection settings.

:::image type="content" source="./media/create-fabric-workspace-finops/fabric-new-shortcut-connection-settings.png" border="true" alt-text="Screenshot showing the New shortcut connection settings." lightbox="./media/create-fabric-workspace-finops/fabric-new-shortcut-connection-settings.png" :::

>[!IMPORTANT]

> >You must have the *Storage Blob Data Contributor* role when you use *Organizational account* authentication. If you don't have the required permission, you can use other authentication types like *Account key*. Otherwise, you get an `Invalid credentials` error.

Reviewer: Should this be listed in prereqs? Or maybe as a step for creating the storage account?

Author: Moved to prereqs

To get the Data Lake Storage URL, view the storage account where the export created a directory and the FOCUS cost file. Under **Settings**, select **Endpoints**. Copy the URL marked as **Data Lake Storage**. 

Here’s an example screenshot showing the Data Lake Storage URL on the **Endpoints** page.

:::image type="content" source="./media/create-fabric-workspace-finops/endpoints-page-storage-account.png" border="true" alt-text="Screenshot showing the Endpoints page of the storage account." lightbox="./media/create-fabric-workspace-finops/endpoints-page-storage-account.png" :::

> ## Copy data into Fabric

Reviewer: ```suggestion
<br>

## Copy data into Fabric
```

1. In the lakehouse, find the directory that you created when you set up the export. The Directory is in the **Files** section.
2. Next to the directory, select the ellipsis (**...**), and then select **Load to Tables** > **New table**.  
    :::image type="content" source="./media/create-fabric-workspace-finops/load-to-tables-new-table.png" border="true" alt-text="Screenshot showing the New table option." lightbox="./media/create-fabric-workspace-finops/load-to-tables-new-table.png" :::
    1. Table name = &lt;*Any valid name*&gt;
    2. File type = `CSV`
    3. Including subfolders = **Enabled**
    4. Use header for column names = **Enabled**
    5. Separator = `,`
    6. Select **Load**

:::image type="content" source="./media/create-fabric-workspace-finops/load-folder-to-new-table.png" border="true" alt-text="Screenshot showing the new table options." lightbox="./media/create-fabric-workspace-finops/load-folder-to-new-table.png" :::

For more information, see [Lakehouse Load to Delta Lake tables](/fabric/data-engineering/load-to-tables).

> This process creates a table based on the CSV/Parquet file and you can transform data using Power Query in the report. For an automated process to ingest data using notebooks, see [Automate data ingestion](#automate-data-ingestion) later in this article.

Reviewer: I don't know if the "you can transform data using Power Query" part is true. We should confirm.

Author: Sure, who can confirm?

Reviewer: Let me see if Joseph knows. Otherwise, I'll have to ask the Fabric team.

Author: Pinged Jojo

Reviewer: ```suggestion
This process creates a table based on the CSV/Parquet file. For an automated process to ingest data using notebooks, see [Automate data ingestion](#automate-data-ingestion) later in this article.
```

> This process creates a table based on the CSV/Parquet file and you can transform data using Power Query in the report. For an automated process to ingest data using notebooks, see [Automate data ingestion](#automate-data-ingestion) later in this article.

Reviewer: I don't think we should show the one-time ingestion since that's not what they'll need to make this useful. Can we replace this with the automated ingestion steps?

Author: @flanakin Please suggest the text change here, if possible.

Reviewer: This is a big change. Probably wouldn't work as a suggestion.

Here’s an example screenshot showing data in the Lakehouse table.

:::image type="content" source="./media/create-fabric-workspace-finops/fabric-load-table-lakehouse.png" border="true" alt-text="Screenshot showing data in the table." lightbox="./media/create-fabric-workspace-finops/fabric-load-table-lakehouse.png" :::

> ## Create a Power BI report

Reviewer: ```suggestion
<br>

## Create a Power BI report
```

> After the data is ingested into Fabric and Delta parquet tables are ready, you can move on to reporting.

Reviewer: No need to mention "delta parquet". They won't know (or need to know) what that means.
```suggestion
After the data is ingested into Fabric and tables are ready, you can move on to reporting.
```

> 1. Create a semantic model by selecting **New semantic model** and naming the model. For more information, see [Default Power BI semantic models](/fabric/data-warehouse/semantic-models#sync-the-default-power-bi-semantic-model).

Reviewer: Should we add where to start?

Author: Yes. Updated.

> 2. Select the Power BI symbol at the bottom left.

Reviewer: We should confirm if this will always be there. I got the impression that icon changes for different contexts and isn't always the same.

Author: Who can confirm?

Author: No action.

3. Select the same semantic model to build reports and then use Copilot to generate insights from your ingested data.

>[!NOTE]
> The Power BI demo uses simulated cost data for illustration purposes.

## Automate data ingestion

For detailed information about automating the process to ingest data into Fabric tables, see [How to use notebooks](/fabric/data-engineering/how-to-use-notebook). 

Here are the high-level steps:

> 1. Open the lakehouse instance.
> 2. Create a notebook as follows:
>     1. Select **Open notebook** > **New notebook**. New notebook might be hidden behind an ellipsis (...) menu overflow symbol.
>     2. Select **PySpark (Python)** notebook and paste the following example Python script.
>     3. Replace `read_path` with &lt;abs-folder-path&gt;. You can find the Azure Blob Filesystem driver (ABFS) path by navigating to the desired path. Select the ellipsis (**...**), and then select **Copy ABFS path** as shown in the following screenshot.  
>         :::image type="content" source="./media/create-fabric-workspace-finops/fabric-lakehouse-copy-path-load-table.png" border="true" alt-text="Screenshot an example of copying ABFS path to load table." lightbox="./media/create-fabric-workspace-finops/fabric-lakehouse-copy-path-load-table.png" :::
>     4. Replace `export_table` with the name for the table that you want to create.
>     5. Running the script ingests the data into Fabric.
> 3. Automate data ingestion
>     1. You can create a schedule to automate running the script that results in automating the data ingestion pipeline.
>     2. To create a daily automation schedule, select the gear symbol on the left menu and select the **Schedule** option.  
>         :::image type="content" source="./media/create-fabric-workspace-finops/schedule-tab.png" border="true" alt-text="Screenshot showing the Schedule tab to automate data ingestion." lightbox="./media/create-fabric-workspace-finops/schedule-tab.png" :::

Reviewer: Is it simpler to flatten the steps?
```suggestion
1. Open the lakehouse instance.
2. Select **Open notebook** > **New notebook**. New notebook might be hidden behind an ellipsis (...) menu overflow symbol.
3. Select **PySpark (Python)** notebook and paste the following example Python script.
4. Replace `read_path` with &lt;abs-folder-path&gt;. You can find the Azure Blob Filesystem driver (ABFS) path by navigating to the desired path. Select the ellipsis (**...**), and then select **Copy ABFS path** as shown in the following screenshot.  
   :::image type="content" source="./media/create-fabric-workspace-finops/fabric-lakehouse-copy-path-load-table.png" border="true" alt-text="Screenshot an example of copying ABFS path to load table." lightbox="./media/create-fabric-workspace-finops/fabric-lakehouse-copy-path-load-table.png" :::
5. Replace `export_table` with the name for the table that you want to create.
6. Running the script ingests the data into Fabric.
7. You can create a schedule to automate running the script that results in automating the data ingestion pipeline.
8. To create a daily automation schedule, select the gear symbol on the left menu and select the **Schedule** option.  
   :::image type="content" source="./media/create-fabric-workspace-finops/schedule-tab.png" border="true" alt-text="Screenshot showing the Schedule tab to automate data ingestion." lightbox="./media/create-fabric-workspace-finops/schedule-tab.png" :::
```

> ## Example Python script

Reviewer: Should this be an h3 so it's under the same section?
```suggestion
### Example Python script
```

> Here's an example Python script to ingest data into Fabric tables that gets used in the preceding [Automate data ingestion](#automate-data-ingestion) section.

Reviewer: ```suggestion
Use the following Python script in step 3 above to ingest data into Fabric tables.
```

```python
#!/usr/bin/env python
# coding: utf-8

# ## Sample Read CSV to Delta Output
# 
# Exports Platform Demo Notebook

# # Export to Delta Parquet

# In[1]:
read_path = '<USER INPUT>' # input path to your data stream
export_table = '<USER INPUT>' # output path for your delta parquet table, ex: "Tables/CostDetails" 

# In[3]:
df = spark.read.format("csv").option("header", "true").option("inferSchema", "true").option("recursiveFileLookup", "true").load(read_path) 
     
''' Use this for parquet reading 
df = spark.read.format("parquet").option("header", "true").option("inferSchema", "true").option("recursiveFileLookup", "true").load(read_path) 
''' 

# In[4]:
''' Use this if you want to overwrite the schema as well 
df.write.mode("overwrite").format("delta").option("overwriteSchema", "true").save(export_table) 
''' 
df.write.mode("overwrite").format("delta").save(export_table) 
```

> ## Related content

Reviewer: ```suggestion
<br>

## Related content
```

> - [FinOps toolkit](https://microsoft.github.io/finops-toolkit/optimization-workbook)

Reviewer: ```suggestion
Related products:

- [Cost Management](/azure/cost-management-billing/costs)

Related solutions:

- [FinOps hubs](https://aka.ms/finops/hubs)
- [FinOps toolkit Power BI reports](https://aka.ms/ftk/pbi)
```

<br>

## Create a Power BI report

After the data is ingested into Fabric and tables are ready, you can move on to reporting.

1. Sign in to the Fabric app using your [Fabric free account](https://www.microsoft.com/microsoft-fabric/getting-started) or [Microsoft Power BI](https://app.powerbi.com) account.
1. Select the workspace where you created the lakehouse and then select the lakehouse.
1. In the Explorer pane, select **Tables** and then select the table that you created.
1. At the top of the page, select **New semantic model** and name the model. For more information, see [Default Power BI semantic models](/fabric/data-warehouse/semantic-models#sync-the-default-power-bi-semantic-model).
4. Select the Power BI symbol at the bottom left.
5. Select the same semantic model to build reports and then use Copilot to generate insights from your ingested data.

>[!NOTE]
> The Power BI demo uses simulated cost data for illustration purposes.

> ## Automate data ingestion

Reviewer: ```suggestion
<br>

## Automate data ingestion
```
